libtransact/src/workload/runner.rs

```diff
@@ -344,6 +344,39 @@ pub struct ServerError {
     pub message: String,
 }
 
+fn slow_rate(
+    target: &str,
+    auth: &str,
+    batch_bytes: Vec<u8>,
+    start_time: time::Instant,
+    submitted_batches: u32,
+) -> Result<(), WorkloadRunnerError> {
+    debug!("Received TooManyRequests message from target, attempting to resubmit batch");
+    // figure out the rate at which batches were successfully submitted
+    let effective_rate = submitted_batches / (time::Instant::now() - start_time).as_secs() as u32;
+    let wait = time::Duration::from_secs(1) / effective_rate;
+    // sleep
+    thread::sleep(wait);
+    loop {
+        // attempt to submit batch again
+        match submit_batch(target, &auth, batch_bytes.clone()) {
+            Ok(()) => break,
+            Err(err) => {
+                // if TooManyRequests is still being returned sleep and repeat loop
+                if err == WorkloadRunnerError::TooManyRequests {
```

Comment:
Instead of this nested if, you could match on this specific error.

Reply:
Updated